File: Algorithms/1. Sorting/merge-sort/mergeSort.py
def merge(S1, S2, S):
  # Merge two sorted Python lists S1 and S2 into properly sized list S
  i = 0; # current location of S1 pointer
  j = 0; # current location of S2 pointer
  while i+j < len(S):
    if j == len(S2) or (i < len(S1) and S1[i] < S2[j]):
      S[i+j] = S1[i] # copy ith element of S1 as next item of S
      i += 1; # increment S1 pointer index
    else:
      S[i+j] = S2[j] # copy jth element of S2 as next item of S
      j += 1; # increment S2 pointer index

def merge_sort(S):
  n = len(S)
  # if the list has only 1 element
  if n < 2:
    return 
  
  # divide the list
  mid = n // 2
  S1 = S[0:mid] # copy of first half
  S2 = S[mid:n] # copy of second half

  # conquer with recursion
  merge_sort(S1) # recursively sort the first half
  merge_sort(S2) # recursively sort the second half

  # merge the 2 sorted list
  merge(S1, S2, S)

def main():
  print("Mergesort O(nlogn) Algorithm")

  # before sorting
  print("Array before sorting:")
  array = [5,4,3,2,1]
  print(array)

  # after sorting
  merge_sort(array)
  print("Array after sorting:")
  print(array)

main();

File: Algorithms/1. Sorting/merge-sort/mergeSort.cpp
// how to run: c++ mergeSort.cpp -std=c++11
#include <iostream>
#include <vector>
using namespace std;

// merge two parts of the sorted array
void merge(vector<int>& array, int low, int mid, int high) {
  // compare the vector and assign it to the temp
  vector<int> temp(array.size(), {0});
  int low1 = low, low2 = mid+1, i = low;
  for (low1, low2, i; low1 <= mid && low2 <= high; i++) {
    if (array[low1] <= array[low2])
      temp[i] = array[low1++];
    else
      temp[i] = array[low2++];
  }

  // assign rest values to the temp vector
  while (low1 <= mid)
    temp[i++] = array[low1++];
  while (low2 <= high)
    temp[i++] = array[low2++];
  
  // copy the sorted array to the original array
  for (i = low; i <= high; i++)
    array[i] = temp[i];
}

// sort the array
void sort(vector<int>& array, int low, int high) {
  if (low < high) {
    int mid = (low + high) / 2; // get mid position
    sort(array, low, mid); // sort the first half
    sort(array, mid+1, high); // sort the second half
    merge(array, low, mid, high); // merge both halves
  }
  else {
    return; // size check
  }
}

int main() {
  cout << "Mergesort O(nlogn) Algorithm\n";
  vector<int> array{9,8,7,6,5,4,3,2,1};
  cout << "Before sorting: ";
  for (auto num: array)
    cout << num << " ";
  cout << endl;

  // run merge sort on given array
  sort(array, 0, array.size()-1);
  cout << "After sorting: ";
  for (auto num: array)
    cout << num << " ";
  cout << endl;
}